### PR TITLE
feat: Enable subsampling of events via configuration

### PR DESCRIPTION
**Pull request by peterychang**

### CODE REVIEW

File: rlclientlib/utility/config_helper.cc [hunk 1]
```diff
@@ -55,6 +68,8 @@ async_batcher_config get_batcher_config(const configuration &config, const char
   res.send_queue_max_capacity = get_int(config, section, name::SEND_QUEUE_MAX_CAPACITY_KB, 16 * 1024) * 1024;
   res.queue_mode = to_queue_mode_enum(get_str(config, section, name::QUEUE_MODE, value::QUEUE_MODE_DROP));
   res.batch_content_encoding = config.get_bool(section, name::USE_DEDUP, false) ? value::CONTENT_ENCODING_DEDUP : value::CONTENT_ENCODING_IDENTITY;
+  // TODO: what should happen if the rate is not (0, 1]?
```
Comment (olgavrou) [inline rlclientlib/utility/config_helper.cc:71]: should we maybe put a warning/throw something here until we figure out what happens if the rate is not (0,1]?

File: unit_test/async_batcher_test.cc [hunk 2]
```diff
@@ -61,6 +64,31 @@ class test_droppable_event : public event {
...
+    return *this;
+  }
+
+  bool try_drop(float drop_prob, int _drop_pass) override {
+    auto prob = static_cast<float>(std::atof(_seed_id.c_str()));
+
+    // logic because floating point comparison is dumb. In this case, 0.7 > 0.7 == true
+    const float tol = 1e-5f;
+    bool is_equal = (prob-drop_prob)*(prob-drop_prob) < tol*tol;
+    return (prob > drop_prob) && !VW::math::are_same(prob, drop_prob, 1e-5f);
```
Comment (jackgerrits) [inline unit_test/async_batcher_test.cc:87]:
```suggestion
    constexpr float FLOAT_TOL= 1e-5f;
    return (prob > drop_prob) && !VW::math::are_same(prob, drop_prob, FLOAT_TOL);
```

File: unit_test/async_batcher_test.cc [hunk 3]
```diff
@@ -84,12 +112,26 @@ namespace reinforcement_learning { namespace logger {
...
+  template <>
+  struct json_event_serializer<config_drop_event> {
+    using serializer_t = json_event_serializer<config_drop_event>;
+
+    static int serialize(config_drop_event& evt, std::ostream& out, api_status* status) {
+      out << evt.get_event_id();
+      return error_code::success;
+    }
+
+    static size_t size_estimate(const config_drop_event& evt) { return 1; }
```
Comment (jackgerrits) [inline unit_test/async_batcher_test.cc:125]: Isn't the size the length of the event id?

File: unit_test/async_batcher_test.cc [hunk 4]
```diff
@@ -61,6 +64,31 @@ class test_droppable_event : public event {
...
+public:
+  config_drop_event() {}
+  config_drop_event(const std::string& id) : event(id.c_str(), timestamp{}) {}
+
+  config_drop_event(config_drop_event&& other) : event(std::move(other)) {}
+  config_drop_event& operator=(config_drop_event&& other)
+  {
+    if (&other != this) event::operator=(std::move(other));
+    return *this;
+  }
```
[the remark this answers is not in the document]
Comment (peterychang) [reply unit_test/async_batcher_test.cc:79]: This was just copied from other test events

File: unit_test/async_batcher_test.cc (hunk 3 above)
Comment (peterychang) [reply unit_test/async_batcher_test.cc:124]: I'm not sure, I just copied it from the previous examples. It probably is? But I assume this estimate is only used to try and optimize the number of memory allocations, I dont think anything will break from it being wrong

File: unit_test/async_batcher_test.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (kumpera) [reply unit_test/async_batcher_test.cc:79]: That doesn't make it any less 🤮

File: unit_test/async_batcher_test.cc (hunk 3 above)
Comment (kumpera) [reply unit_test/async_batcher_test.cc:124]: Size estimate can be anything so any positive value should work.

File: unit_test/async_batcher_test.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (jackgerrits) [reply unit_test/async_batcher_test.cc:79]: Then perhaps it should be removed from the others too?

**Pull request merged**